File: chunkscan/src/client/java/dev/jack/chunkscan/Fleet.java
package dev.jack.chunkscan;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import net.minecraft.client.Minecraft;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.time.LocalDateTime;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Which of this island's designs THIS account is building.
 *
 * <p>Jack runs up to five alts and they all read the same schematics folder on one machine, which
 * removes almost everything a fleet would otherwise have to solve: no per-account profile, no
 * separate design set, nothing to sync. Every client already sees every design. What is left is
 * the only hard part — deciding who builds what — and it is one shared file, {@code fleet.json},
 * written by {@code python -m mcbuild fleet} and claimed from in game.
 *
 * <p><b>TWO ACCOUNTS MUST NEVER SHARE A DESIGN.</b> Not because the files collide — they are the
 * same file — but because two printers placing into the same cells fight: each one sees the
 * other's block as either already built or as a deviation, and {@link Printer}'s report, the one
 * honest signal this loop has, becomes noise. So a design is claimed exclusively.
 *
 * <p><b>A CLAIM IS A LEASE, BECAUSE AN ALT CAN DIE.</b> A client crashes, a session drops, someone
 * logs out mid-build. A claim that never expires strands a design for ever; one with no expiry at
 * all lets a second account grab it the moment the first is slow. So it is refreshed by a
 * heartbeat while the loop runs, and swept when it goes quiet — the same reasoning the chest
 * cooling-off already carries, and it has to expire for exactly the same reason.
 *
 * <p><b>FIVE IS A SERVER RULE.</b> It is enforced on the Python side where the assignment is made;
 * this side reports it so a sixth client cannot quietly appear without anyone seeing.
 */
final class Fleet {
	static final String FILE = "fleet.json";
	/** Matches {@code fleet.LEASE_MINUTES}. A claim unrefreshed for this long is handed back. */
	static final int LEASE_MINUTES = 15;
	/** How often the loop refreshes its claims. Well under the lease, so a busy client never loses one. */
	static final long BEAT_MS = 60_000;

	static final class Claim {
		String account = "";
		String at = "";
		String seen = "";
	}

	static final class State {
		String plan = "";
		Map<String, Claim> claims = new LinkedHashMap<>();
		List<String> done = new ArrayList<>();
	}

	private static long lastBeat;

	private Fleet() {}

	/** This client's account name — whoever is logged in. There is nothing else to configure. */
	static String me(Minecraft mc) {
		return mc.getUser() == null ? "" : mc.getUser().getName();
	}

	static State load(Path dir) {
		Path f = dir.resolve(FILE);
		if (!Files.exists(f)) return new State();
		try {
			JsonObject root = JsonParser.parseString(Files.readString(f, StandardCharsets.UTF_8))
				.getAsJsonObject();
			State s = new Gson().fromJson(root, State.class);
			if (s == null) return new State();
			if (s.claims == null) s.claims = new LinkedHashMap<>();
			if (s.done == null) s.done = new ArrayList<>();
			return s;
		} catch (Exception e) {
			ChunkScanClient.LOG.warn("fleet.json unreadable: {}", e.toString());
			return new State();
		}
	}

	/**
	 * Write atomically: temp file, then move.
	 *
	 * <p><b>FIVE CLIENTS SHARE THIS FILE.</b> A plain write leaves a window in which the file on
	 * disk is truncated, and a client reading in that window sees an EMPTY fleet — and cheerfully
	 * claims a design another account is already building, which is the one thing this file exists
	 * to prevent. An atomic move means a reader sees the old file or the new one, never half of one.
	 *
	 * <p>It does not make read-modify-write atomic, and pretending otherwise would be worse than
	 * the bug. Two clients that both read and then both write still lose one change. What carries
	 * it is that claims are near-idempotent and heartbeats repeat: a lost heartbeat is retried a
	 * minute later and a lost claim shows up in the next report. A real lock is the fix the day
	 * this assigns something that cannot be repeated.
	 */
	static void save(Path dir, State s) throws Exception {
		Files.createDirectories(dir);
		Path tmp = dir.resolve(FILE + ".tmp");
		Files.writeString(tmp, new GsonBuilder().setPrettyPrinting().create().toJson(s),
			StandardCharsets.UTF_8);
		try {
			Files.move(tmp, dir.resolve(FILE), java.nio.file.StandardCopyOption.REPLACE_EXISTING,
				java.nio.file.StandardCopyOption.ATOMIC_MOVE);
		} catch (java.nio.file.AtomicMoveNotSupportedException e) {
			Files.move(tmp, dir.resolve(FILE), java.nio.file.StandardCopyOption.REPLACE_EXISTING);
		}
	}

	static boolean expired(Claim c) {
		try {
			String seen = (c.seen == null || c.seen.isBlank()) ? c.at : c.seen;
			return Duration.between(LocalDateTime.parse(seen), LocalDateTime.now()).toMinutes()
				>= LEASE_MINUTES;
		} catch (DateTimeParseException | NullPointerException e) {
			return false;      // an unparseable stamp is not evidence the account is gone
		}
	}

	/** Designs this account holds, expired claims excluded. */
	static List<String> mine(Path dir, String account) {
		List<String> out = new ArrayList<>();
		State s = load(dir);
		for (var e : s.claims.entrySet()) {
			if (e.getValue().account.equals(account) && !expired(e.getValue())) out.add(e.getKey());
		}
		return out;
	}

	/** True when somebody ELSE holds this design and their lease is still good. */
	static boolean heldByOther(Path dir, String design, String account) {
		Claim c = load(dir).claims.get(design);
		return c != null && !c.account.equals(account) && !expired(c);
	}

	static String claim(Path dir, String design, String account) throws Exception {
		State s = load(dir);
		Claim cur = s.claims.get(design);
		if (cur != null && !cur.account.equals(account) && !expired(cur)) {
			return design + " is held by " + cur.account + " - two printers on one design fight";
		}
		Claim c = new Claim();
		c.account = account;
		c.at = LocalDateTime.now().toString();
		c.seen = c.at;
		s.claims.put(design, c);
		save(dir, s);
		return account + " holds " + design;
	}

	/** Refresh every claim this account holds. Rate-limited: the loop calls it every tick. */
	static void heartbeat(Minecraft mc) {
		if (System.currentTimeMillis() - lastBeat < BEAT_MS) return;
		lastBeat = System.currentTimeMillis();
		try {
			Path dir = ScanRunner.schematicsDir(mc);
			String me = me(mc);
			State s = load(dir);
			boolean any = false;
			for (Claim c : s.claims.values()) {
				if (c.account.equals(me)) {
					c.seen = LocalDateTime.now().toString();
					any = true;
				}
			}
			if (any) save(dir, s);
		} catch (Exception e) {
			ChunkScanClient.LOG.warn("fleet heartbeat failed: {}", e.toString());
		}
	}

	static String finish(Path dir, String design, String account) throws Exception {
		State s = load(dir);
		s.claims.remove(design);
		if (!s.done.contains(design)) s.done.add(design);
		save(dir, s);
		return account + " finished " + design;
	}

	static String report(Path dir, String me) {
		State s = load(dir);
		StringBuilder b = new StringBuilder();
		b.append("fleet: ").append(s.claims.size()).append(" claimed, ")
			.append(s.done.size()).append(" done");
		if (s.plan != null && !s.plan.isBlank()) b.append("  plan ").append(s.plan);
		b.append("\n  you are ").append(me);
		List<String> mine = mine(dir, me);
		b.append(mine.isEmpty() ? " and hold nothing" : " and hold: " + String.join(", ", mine));
		for (var e : s.claims.entrySet()) {
			if (e.getValue().account.equals(me)) continue;
			b.append("\n  ").append(e.getValue().account).append(": ").append(e.getKey())
				.append(expired(e.getValue()) ? "  (LEASE EXPIRED - free to take)" : "");
		}
		return b.toString();
	}
}

File: mcbuild/fleet.py
"""Up to five accounts building at once, out of ONE shared schematics folder.

    python -m mcbuild fleet --assign casino          # split an approved plan across the alts
    python -m mcbuild fleet                          # who has what, and what is still unclaimed
    python -m mcbuild fleet --release Enroniti3      # hand a stalled account's work back

Jack runs up to five alts, which he has confirmed is what the server allows, and they all read the
same schematics directory on one machine. **That simplification removes the whole problem this file
would otherwise be about.** There is no per-account profile, no separate design set, no syncing:
every client already sees every design. What is left is the only genuinely hard part, which is
deciding WHO BUILDS WHAT.

**FIVE IS A CAP, NOT A DEFAULT.** It is a server rule, so it is enforced here rather than written
in a comment, and `MAX_ACCOUNTS` is checked on every assignment. A tool that quietly allowed a
sixth would be a tool that got the account banned.

**AND TWO ACCOUNTS MUST NEVER SHARE A DESIGN.** Not because the schematics collide - they cannot,
they are the same file - but because two printers placing into the same cells fight: each sees the
other's block as "already built" or as a deviation, and the loop's whole feedback signal becomes
noise. The unit of assignment is therefore a DESIGN, claimed exclusively, with a lease.

**A LEASE, BECAUSE AN ALT CAN DIE.** A client crashes, a session drops, someone logs out mid-build.
A claim that never expires strands a design for ever; a claim with no expiry at all lets two
accounts pick it up the moment one of them is slow. `LEASE_MINUTES` is how long a claim survives
without a heartbeat, and `sweep()` is what hands the work back - the same reasoning as the chest
cooling-off in the build loop, which had to expire for exactly the same reason.
"""
from __future__ import annotations

import datetime as _dt
import json
import os
import pathlib

MAX_ACCOUNTS = 5                 # skyblock.net's rule, enforced rather than remembered
LEASE_MINUTES = 15
FILE = "fleet.json"


def path(schem_dir: str | None = None) -> str:
    from .profile import load as load_profile
    return os.path.join(schem_dir or load_profile()["schem_dir"], FILE)


def _now() -> _dt.datetime:
    return _dt.datetime.now()


def _key(island: str | None, design: str) -> str:
    """A claim is keyed by ISLAND AND DESIGN.

    Two alts can be building designs of the same name on two different islands, and the moment
    that is true a bare design name is ambiguous - the claim would be handed to whichever asked
    first and the other account would be told its own work was taken. The island comes from the
    design's own sidecar, so nothing has to be told where it is.
    """
    return f"{island}/{design}" if island else design


def load(schem_dir: str | None = None) -> dict:
    p = path(schem_dir)
    if not os.path.exists(p):
        return {"plan": "", "claims": {}, "done": []}
    try:
        with open(p, encoding="utf-8") as f:
            d = json.load(f) or {}
    except (OSError, ValueError):
        return {"plan": "", "claims": {}, "done": []}
    d.setdefault("claims", {})
    d.setdefault("done", [])
    return d


def save(state: dict, schem_dir: str | None = None) -> str:
    """Write atomically: temp file, then replace.

    FIVE CLIENTS SHARE THIS FILE. A plain open-write leaves a window where the file on disk is
    truncated, and a client that reads in that window sees an empty fleet and cheerfully claims
    a design somebody else is already building. `os.replace` is atomic on Windows and POSIX, so a
    reader sees either the old file or the new one and never a half-written one.

    THIS DOES NOT MAKE READ-MODIFY-WRITE ATOMIC, and pretending otherwise would be worse than the
    bug: two clients that both read, then both write, still lose one of the two changes. What
    saves it in practice is that claims are near-idempotent and heartbeats repeat - a lost
    heartbeat is retried a minute later, and a lost claim is noticed by the next `report`. A real
    lock is the fix if this ever assigns anything that cannot be repeated.
    """
    p = path(schem_dir)
    pathlib.Path(p).parent.mkdir(parents=True, exist_ok=True)
    tmp = p + ".tmp"
    with open(tmp, "w", encoding="utf-8") as f:
        json.dump(state, f, indent=1)
        f.flush()
        os.fsync(f.fileno())
    os.replace(tmp, p)
    return p


def sweep(state: dict) -> list:
    """Drop claims whose lease has run out. Returns the designs handed back.

    UNLOADED IS NOT ABSENT, and neither is quiet: an account that has not checked in may be
    mid-flight across the island. The lease is generous for that reason - it exists to recover from
    a dead client, not to punish a slow one.
    """
    freed = []
    for design, claim in list(state["claims"].items()):
        try:
            seen = _dt.datetime.fromisoformat(claim.get("seen", claim.get("at", "")))
        except ValueError:
            continue
        if (_now() - seen).total_seconds() > LEASE_MINUTES * 60:
            freed.append(design)
            state["claims"].pop(design)
    return freed


def assign(designs: list, accounts: list, schem_dir: str | None = None,
           island: str | None = None, islands_of: dict | None = None) -> dict:
    """Hand each account a share of the work. Existing live claims are respected.

    `islands_of` maps design -> island when the designs are spread over more than one; `island`
    forces them all onto one. Neither is required, and with neither the behaviour is exactly what
    it was before there were two islands.
    """
    islands_of = islands_of or {}
    accounts = [a for a in accounts if a]
    if len(accounts) > MAX_ACCOUNTS:
        raise ValueError(f"{len(accounts)} accounts, and the cap is {MAX_ACCOUNTS} - "
                         f"that is a server rule, not a setting")
    if not accounts:
        raise ValueError("no accounts given")
    state = load(schem_dir)
    sweep(state)
    held = {c["account"] for c in state["claims"].values()}
    free = [d for d in designs
            if _key(island or islands_of.get(d, ""), d) not in state["claims"]
            and d not in state["done"]]
    # Round-robin over the accounts that are actually in this fleet, so an account that already
    # holds something is not handed a second design before the idle ones have one.
    order = sorted(accounts, key=lambda a: (a in held, a))
    for i, design in enumerate(free):
        acct = order[i % len(order)]
        isl = island or islands_of.get(design, "")
        state["claims"][_key(isl, design)] = {
            "account": acct, "design": design, "island": isl,
            "at": _now().isoformat(timespec="seconds"),
            "seen": _now().isoformat(timespec="seconds")}
    save(state, schem_dir)
    return state


def claim(design: str, account: str, schem_dir: str | None = None,
          island: str | None = None) -> tuple:
    """One account taking one design ON ONE ISLAND. (ok, why); refuses what somebody else holds."""
    state = load(schem_dir)
    sweep(state)
    k = _key(island, design)
    cur = state["claims"].get(k)
    if cur and cur["account"] != account:
        return False, f"{k} is held by {cur['account']}"
    state["claims"][k] = {"account": account, "design": design, "island": island or "",
                          "at": _now().isoformat(timespec="seconds"),
                          "seen": _now().isoformat(timespec="seconds")}
    save(state, schem_dir)
    return True, f"{account} holds {k}"


def heartbeat(account: str, schem_dir: str | None = None) -> int:
    """Refresh every claim this account holds. Returns how many."""
    state = load(schem_dir)
    n = 0
    for claim_ in state["claims"].values():
        if claim_["account"] == account:
            claim_["seen"] = _now().isoformat(timespec="seconds")
            n += 1
    save(state, schem_dir)
    return n


def finish(design: str, account: str, schem_dir: str | None = None) -> str:
    state = load(schem_dir)
    state["claims"].pop(design, None)
    if design not in state["done"]:
        state["done"].append(design)
    save(state, schem_dir)
    return f"{account} finished {design}"


def release(account: str, schem_dir: str | None = None) -> list:
    state = load(schem_dir)
    freed = [d for d, c in state["claims"].items() if c["account"] == account]
    for d in freed:
        state["claims"].pop(d)
    save(state, schem_dir)
    return freed


def mine(account: str, schem_dir: str | None = None, island: str | None = None) -> list:
    """What this account holds — everywhere, or on one island."""
    state = load(schem_dir)
    sweep(state)
    out = []
    for k, c in state["claims"].items():
        if c["account"] != account:
            continue
        if island is not None and c.get("island", "") != island:
            continue
        out.append(c.get("design", k))
    return out


def report(schem_dir: str | None = None) -> str:
    state = load(schem_dir)
    freed = sweep(state)
    if freed:
        save(state, schem_dir)
    lines = [f"fleet: {len(state['claims'])} design(s) claimed, {len(state['done'])} done"
             f"  (cap {MAX_ACCOUNTS} accounts, lease {LEASE_MINUTES} min)"]
    if state.get("plan"):
        lines.append(f"  plan: {state['plan']}")
    by: dict = {}
    for d, c in state["claims"].items():
        by.setdefault(c["account"], []).append((d, c.get("seen", "")))
    for acct in sorted(by):
        lines.append(f"  {acct}:")
        for d, seen in sorted(by[acct]):
            age = ""
            try:
                mins = int((_now() - _dt.datetime.fromisoformat(seen)).total_seconds() // 60)
                age = f"  (last seen {mins} min ago)" if mins else "  (just now)"
            except ValueError:
                pass
            lines.append(f"    {d}{age}")
    if freed:
        lines.append(f"  lease expired, handed back: {', '.join(freed)}")
    if state["done"]:
        lines.append("  done: " + ", ".join(state["done"][:8])
                     + (f" ...and {len(state['done']) - 8} more" if len(state["done"]) > 8 else ""))
    if not state["claims"] and not state["done"]:
        lines.append("  nothing assigned - python -m mcbuild fleet --assign <plan>")
    return "\n".join(lines)
